### PR TITLE
Fix Mdns, HTTPS server for `esp-idf` v5.0+, option for httpd wildcard matcher (available in v4 as well)

### PR DESCRIPTION
**Pull request by zRedShift**

### CODE REVIEW

File: src/mdns.rs [hunk 1]
```diff
@@ -88,6 +88,17 @@ impl From<mdns_result_t> for QueryResult {
...
         }
 
+        #[cfg(esp_idf_version_major = "5")]
+        let interface = match unsafe { CStr::from_ptr(esp_netif_get_desc(result.esp_netif)) }
+            .to_bytes_with_nul()
+        {
+            b"sta\0" => Interface::STA,
+            b"ap\0" => Interface::AP,
+            b"eth\0" => Interface::ETH,
+            _ => unimplemented!(),
```
Comment (zRedShift) [inline src/mdns.rs:98]: This is not very elegant, as the user can give custom descriptions to `esp_netif`, and also there are multiple new `netif` types that aren't covered by these.

File: src/mdns.rs [hunk 2]
```diff
@@ -88,6 +88,17 @@ impl From<mdns_result_t> for QueryResult {
             p = unsafe { (*p).next };
         }
 
+        #[cfg(esp_idf_version_major = "5")]
```
Comment (zRedShift) [inline src/mdns.rs:91]: `tcpip_if` -> `esp_netif` in v5.0+
see: https://github.com/espressif/esp-protocols/commit/ddc58e8220f8b4d8829cfe47cfecdf071e92c845#diff-b1f0aa635ad5c4ccef42c50cda8b8e441e9fc5b324d1ea3f7703af1ded32cb75L86-R79
maybe a better approach would be to get rid of the `Interface` enum

File: src/lib.rs [hunk 3]
```diff
@@ -42,7 +42,7 @@ pub mod http;
 #[cfg(all(feature = "std", esp_idf_comp_esp_http_server_enabled))]
 pub mod httpd;
 pub mod log;
-#[cfg(feature = "alloc")]
+#[cfg(all(feature = "alloc", esp_idf_comp_mdns_enabled))]
```
Comment (zRedShift) [inline src/lib.rs:45]: let's leave the work of making sure that the `mdns` component is fetched to `esp-idf-sys`, no need to check `espidf` version here

File: src/http/server.rs [hunk 4]
```diff
@@ -559,9 +578,7 @@ impl<'a> EspHttpConnection<'a> {
     pub fn uri(&self) -> &str {
         self.assert_request();
 
-        let c_uri = unsafe {
-            CStr::from_bytes_with_nul_unchecked(mem::transmute(self.request.0.uri.as_slice()))
-        };
+        let c_uri = unsafe { CStr::from_ptr(self.request.0.uri.as_ptr()) };
```
Comment (zRedShift) [inline src/http/server.rs:581]: This was a bug, it violated the `CStr` safety contract, and ended up as a 512 byte long mostly-zeroed `&str`

File: src/http/server.rs [hunk 5]
```diff
@@ -159,15 +161,23 @@ impl From<&Configuration> for Newtype<httpd_ssl_config_t> {
...
             cacert_pem: ptr::null(),
             cacert_len: 0,
             prvtkey_pem: ptr::null(),
             prvtkey_len: 0,
+            #[cfg(esp_idf_version_major = "4")]
             client_verify_cert_pem: ptr::null(),
+            #[cfg(esp_idf_version_major = "4")]
             client_verify_cert_len: 0,
+            #[cfg(esp_idf_version_major = "5")]
+            servercert: ptr::null(),
```
Comment (zRedShift) [inline src/http/server.rs:178]: renamed in v5.0+
see: https://github.com/espressif/esp-idf/commit/94056fd4a547b63521f9b47fd313b8d1b8a1c7c5#diff-4ac40d704acb6c3ed7b266a6fe23a62209bd770ba94b340d5343381e45980e34L53-L68

File: src/http/server.rs [hunk 6]
```diff
@@ -159,15 +161,23 @@ impl From<&Configuration> for Newtype<httpd_ssl_config_t> {
         Self(httpd_ssl_config_t {
             httpd: http_config.0,
             session_tickets: false,
+            #[cfg(esp_idf_version_major = "5")]
```
Comment (ivmarkov) [inline src/http/server.rs:164]: Please do `not(esp_idf_version_major = "4")`. This way we are safe for ESP-IDF 6 :)

File: src/http/server.rs [hunk 7]
```diff
@@ -159,15 +161,23 @@ impl From<&Configuration> for Newtype<httpd_ssl_config_t> {
...
             transport_mode,
             cacert_pem: ptr::null(),
             cacert_len: 0,
             prvtkey_pem: ptr::null(),
             prvtkey_len: 0,
+            #[cfg(esp_idf_version_major = "4")]
             client_verify_cert_pem: ptr::null(),
+            #[cfg(esp_idf_version_major = "4")]
             client_verify_cert_len: 0,
+            #[cfg(esp_idf_version_major = "5")]
```
Comment (ivmarkov) [inline src/http/server.rs:177]: Ditto.

File: src/http/server.rs [hunk 8]
```diff
@@ -159,15 +161,23 @@ impl From<&Configuration> for Newtype<httpd_ssl_config_t> {
...
+            servercert: ptr::null(),
+            #[cfg(esp_idf_version_major = "5")]
```
Comment (ivmarkov) [inline src/http/server.rs:179]: Ditto.

File: src/http/server.rs [hunk 9]
```diff
@@ -249,8 +259,17 @@ impl EspHttpServer {
...
 
-                config.0.cacert_pem = cert.as_ptr();
-                config.0.cacert_len = cert.len() as u32;
+                #[cfg(esp_idf_version_major = "4")]
+                {
+                    config.0.client_verify_cert_pem = cert.as_ptr();
+                    config.0.servercert_len = cert.len() as u32;
+                }
+
+                #[cfg(esp_idf_version_major = "5")]
```
Comment (ivmarkov) [inline src/http/server.rs:268]: Ditto.

File: src/http/server.rs (hunk 4 above)
Comment (ivmarkov) [reply src/http/server.rs:581]: OK - thanks for fixing!

File: src/lib.rs (hunk 3 above)
Comment (ivmarkov) [reply src/lib.rs:45]: But do we have `esp_idf_comp_mdns_enabled` as a valid cfg in ESP-IDF V4.4? I think so, but if you could confirm. I don't want that we are ending with regressions for V4.4.

File: src/mdns.rs (hunk 2 above)
Comment (ivmarkov) [inline src/mdns.rs:91]: !v4 :)

File: src/mdns.rs (hunk 1 above)
Comment (ivmarkov) [reply src/mdns.rs:98]: Very shaky. I would put a TODO here even though we won't fix it right now.

**Review by ivmarkov — CHANGES_REQUESTED** [review]
If you can address the feedback. Almost all if not all of it is to switch the ESP IDF comparisons to not explicitly reference version "5", as we do everywhere else.

File: src/lib.rs (hunk 3 above)
Comment (zRedShift) [reply src/lib.rs:45]: `esp_idf_comp_mdns_enabled` is a valid cfg in v4.4

**Pull request merged**